librz/bin/dwarf.c

```diff
@@ -1745,278 +1851,237 @@ static const ut8 *parse_die(const ut8 *buf, const ut8 *buf_end, RzBinDwarfDebugI
-		die->offset += unit->hdr.is_64bit ? 12 : 4;
-
+static bool comp_unit_die_parse(RzBuffer *buffer, RzBinDwarfCompUnit *unit, RzBinDwarfDebugInfo *info, const RzBinDwarfAbbrevTable *tbl, RzBuffer *str_buffer, bool big_endian) {
+
+	size_t index = 0;
+	st64 depth = 0;
+	while (true) {
+		ut64 offset = rz_buf_tell(buffer);
+		RzBinDwarfDie die = {
+			.offset = offset,
+			.unit_offset = unit->offset,
+			.index = index,
+			.depth = depth,
+		};
 		// DIE starts with ULEB128 with the abbreviation code
-		ut64 abbr_code;
-		buf = rz_uleb128(buf, buf_end - buf, &abbr_code, NULL);
-
-		if (abbr_code > abbrevs->count || !buf) { // something invalid
-			return NULL;
+		// we wanna store this entry too, usually the last one is null_entry
+		// return the buffer to parse next compilation units
+		if (rz_buf_uleb128(buffer, &die.abbrev_code) < 0) {
+			rz_buf_seek(buffer, (st64)offset, SEEK_SET);
+			break;
 		}
 
-		if (buf >= buf_end) {
-			unit->count++; // we wanna store this entry too, usually the last one is null_entry
-			return buf; // return the buffer to parse next compilation units
-		}
 		// there can be "null" entries that have abbr_code == 0
-		if (!abbr_code) {
-			unit->count++;
-			continue;
+		if (!die.abbrev_code) {
+			RZ_LOG_DEBUG("0x%" PFMT64x ":\t%sNULL\n", offset, indent_str(die.depth));
+			rz_vector_push(&unit->dies, &die);
+			depth--;
+			if (depth <= 0) {
+				break;
+			} else {
+				continue;
+			}
 		}
-		ut64 abbr_idx = first_abbr_idx + abbr_code;
 
-		if (abbrevs->count < abbr_idx) {
-			return NULL;
+		RzBinDwarfAbbrevDecl *abbrev = rz_bin_dwarf_abbrev_get(tbl, die.abbrev_code);
+		if (!abbrev) {
+			break;
 		}
-		RzBinDwarfAbbrevDecl *abbrev = &abbrevs->decls[abbr_idx - 1];
 
-		if (init_die(die, abbr_code, abbrev->count)) {
-			return NULL; // error
+		ut64 attr_count = rz_bin_dwarf_abbrev_decl_count(abbrev);
+		if (attr_count) {
+			rz_vector_init(&die.attrs, sizeof(RzBinDwarfAttr), (RzVectorFree)attr_fini, NULL);
+			rz_vector_reserve(&die.attrs, attr_count);
 		}
-		die->tag = abbrev->tag;
-		die->has_children = abbrev->has_children;
-
-		buf = parse_die(buf, buf_end, info, abbrev, &unit->hdr, die, debug_str, debug_str_len, big_endian);
-		if (!buf) {
-			return NULL;
+		if (abbrev->code != 0) {
+			die.tag = abbrev->tag;
+			die.has_children = abbrev->has_children;
+			if (die.has_children) {
+				depth++;
+			}
+			GOTO_IF_FAIL(die_attr_parse(buffer, &die, info, abbrev, &unit->hdr, str_buffer, big_endian), err);
 		}
-
-		unit->count++;
+		index++;
+		rz_vector_push(&unit->dies, &die);
 	}
-	return buf;
+	return true;
+err:
+	return false;
 }
 
 /**
- * @brief Reads all information about compilation unit header
- *
- * @param buf Start of the buffer
- * @param buf_end Upper bound of the buffer
- * @param unit Unit to read information into
- * @return ut8* Advanced position in a buffer
+ * \brief Reads all information about compilation unit header
```

Comment:
We should add a simple linter to check that Doxygen uses `\syntax` instead of `@syntax`

Reply:
See https://github.com/rizinorg/rizin/issues/3572